pkg/validation/sessions.go

```diff
@@ -35,6 +35,16 @@ func validateSessionCookieMinimal(o *options.Options) []string {
 	return msgs
 }
 
+func validateDynamoDbSessionStore(o *options.Options) []string {
```

Comment:
Once we start implementing testing -- what's the landscape look like for a DynamoDB mocking library that we can use in unit tests?

Reply:
I'll have to take a look on the Go side of things. Coming from a .NET background I've [typically used the Dynamo local JAR binary](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.DownloadingAndRunning.html) for integration style tests to great success.

Reply:
@NickMeves Hi Nick, it looks like Dynamo [provides an interface](https://github.com/aws/aws-sdk-go/blob/v1.35.19/service/dynamodb/dynamodbiface/interface.go#L62-L247) that can be used for stubbing behaviours.

Can you suggest any patterns for replacing the [dynamo service ](https://github.com/oauth2-proxy/oauth2-proxy/blob/3366e9ca0cc703405625c427d7631476769b73ca/pkg/sessions/dynamodb/dynamodb_store.go#L21) from my test? I was considering adding the interface (or a subset of it) to the `NewDynamoDBSessionStore` method so the caller can supply the dependency, but I noticed there are other patterns in Go such as [Monkey Patching](https://medium.com/cbi-engineering/mocking-techniques-for-go-805c10f1676b)?

Reply:
I think emulating the Redis store design is likely your best bet.  It uses the `miniredis` and `minisentinel` libraries to mock a redis client.  Sounds like you have something similar with the stubbing interface you found?

...sounds like you need to abstract the DynamoDB client a bit to match Redis like you originally had...

Sorry for the back and forth if that's the case 😅

Reply:
I'm not sure we do need to do that, the `SessionStore` you've implemented has the interface as a member by the looks of it? Where `NewDynamoDBSessionStore` is created, we can type assert it and override the client with a fake client that we can implement to replace the DynamoDB interface.

Or the alternative is that we create our own interface that is a subset as @JosephWoodward suggested and just create a fake for that, either way, using a type assertion is probably the easiest way to override the client.

This would fall down of course if there's any start up checking that `NewDynamoDBSessionStore` does that would cause it to panic if it can't initialise a client, do we know if that's the case?

Reply:
Hi @JoelSpeed , any thoughts on [this](https://github.com/oauth2-proxy/oauth2-proxy/pull/876/files#diff-bcad1b7019588a2cce9818096233484bd2b7ccbd1a8eae77573855c66c9ec92d) as a solution?

Reply:
Apologies for the delay in getting back to you, yes, I think something like that will work (the link didn't work but I think you're talking about the stuff in `pkg/sessions/dynamodb/dynamodb_store_test.go ` and the GetPutDelete interface), we just need to keep our own table and entries and be able to load and restore the information as the test suite required.

Eventually I imagine we will run the test suite like we do in the Redis package
https://github.com/oauth2-proxy/oauth2-proxy/blob/8bed7aafbdb2cb6241b9655e983dd54d54b50531/pkg/sessions/redis/redis_store_test.go#L74-L89

The first argument to that function will allow you to initialise your fake dynamo client and session store based on that client.

It's been a while since I caught up on this, but does Dynamo have expiry for keys? If so, we should implement something to mock this in the fake store itself, then we pass the fastForwardFunc as the second argument to the test which allows the test to fast forward time to expire the keys earlier

Reply:
> It's been a while since I caught up on this, but does Dynamo have expiry for keys?

Kind of. It has support for a TTL type, however it's not exact and can take up 48 hours to clean up, so in my case I'm storing the expiry as a unix timestamp and comparing against that.